compiler/rustc_trait_selection/src/traits/outlives_bounds.rs

```diff
@@ -55,7 +55,16 @@ impl<'a, 'tcx: 'a> InferCtxtExt<'a, 'tcx> for InferCtxt<'tcx> {
     ) -> Vec<OutlivesBound<'tcx>> {
         let ty = self.resolve_vars_if_possible(ty);
         let ty = OpportunisticRegionResolver::new(self).fold_ty(ty);
-        assert!(!ty.needs_infer());
+
+        // We must avoid processing constrained lifetime variables in implied
+        // bounds. See #110161 for context.
+        if ty.needs_infer() {
+            self.tcx.sess.delay_span_bug(
+                self.tcx.source_span_untracked(body_id),
```

Comment:
Avoid using untracked methods unless necesary.

```suggestion
                self.tcx.def_span(body_id),
```